runtime/dart_isolate.cc

```diff
@@ -681,29 +682,32 @@ bool DartIsolate::DartIsolateInitializeCallback(
       Dart_CurrentIsolateGroupData());
 
   TaskRunners null_task_runners((*raw_embedder_isolate)->GetAdvisoryScriptURI(),
-      /* platform= */ nullptr, /* gpu= */ nullptr, /* ui= */ nullptr,
-      /* io= */ nullptr);
+                                /* platform= */ nullptr, /* gpu= */ nullptr,
+                                /* ui= */ nullptr,
+                                /* io= */ nullptr);
 
   std::unique_ptr<std::shared_ptr<DartIsolate>> embedder_isolate =
-      std::make_unique<std::shared_ptr<DartIsolate>>(
-          std::make_shared<DartIsolate>(
-              (*raw_embedder_isolate)->GetSettings(),         // settings
-              (*raw_embedder_isolate)->GetIsolateSnapshot(),  // isolate_snapshot
-              (*raw_embedder_isolate)->GetSharedSnapshot(),   // shared_snapshot
-              null_task_runners,                              // task_runners
-              fml::WeakPtr<IOManager>{},                      // io_manager
-              fml::WeakPtr<ImageDecoder>{},                   // io_manager
-              (*raw_embedder_isolate)->GetAdvisoryScriptURI(),         // advisory_script_uri
-              (*raw_embedder_isolate)->GetAdvisoryScriptEntrypoint(),  // advisory_script_entrypoint
-              (*raw_embedder_isolate)->child_isolate_preparer_,    // preparer
-              (*raw_embedder_isolate)->isolate_create_callback_,   // on create
-              (*raw_embedder_isolate)->isolate_shutdown_callback_  // on shutdown
-          )
-      );
+      std::make_unique<
+          std::shared_ptr<DartIsolate>>(std::make_shared<DartIsolate>(
```

Comment:
This makes a new `DartIsolate` object, which will need to be freed at some point. I assume the lifetime of this object is limited to the lifetime of *this particular* isolate.

If so, we need to add a isolate-specific cleanup callback in `Dart_Initialize`, namely:

    Dart_InitializeParams params = {};
    ...
    params.cleanup_isolate = ...

Though we have to be careful to not free the `std::shared_ptr<DartIsolate>` object of the root / ui isolate, because this pointer is used as the isolate group data and therefore has to stay alive as long as there are any isolates.

We should ensure isolate creation, deletion cycles work, even in asan (I guess flutter has an asan bot?)

Reply:
> If so, we need to add a isolate-specific cleanup callback in Dart_Initialize

Done

> We should ensure isolate creation, deletion cycles work, even in asan (I guess flutter has an asan bot?)

I don't think flutter has asan bots: these are the bots used by flutter engine: https://ci.chromium.org/p/flutter/g/engine/console